File: src/main/java/com/baedal/store/application/service/StoreService.java
package com.baedal.store.application.service;

import com.baedal.store.application.command.AddStoreCommand;
import com.baedal.store.application.command.DeliveryInfoCommand;
import com.baedal.store.application.command.GetStoreDetailCommand;
import com.baedal.store.application.command.SearchNameCommand.Request;
import com.baedal.store.application.command.SearchNameCommand.Response;
import com.baedal.store.application.command.ValidateOrderInfoCommand;
import com.baedal.store.application.mapper.StoreApplicationMapper;
import com.baedal.store.application.port.in.StoreUseCase;
import com.baedal.store.application.port.out.MessageSenderPort;
import com.baedal.store.application.port.out.ProductPort;
import com.baedal.store.application.port.out.ReviewPort;
import com.baedal.store.application.port.out.StoreRepositoryPort;
import com.baedal.store.application.port.out.StoreSearchRepositoryPort;
import com.baedal.store.domain.business.StoreValidator;
import com.baedal.store.domain.business.VirtualThreadManager;
import com.baedal.store.domain.model.ProductInfo;
import com.baedal.store.domain.model.Store;
import com.baedal.store.domain.model.StoreReviewSummary;
import com.baedal.store.util.StructuredTaskUtil;
import java.util.List;
import java.util.concurrent.StructuredTaskScope.ShutdownOnFailure;
import java.util.concurrent.StructuredTaskScope.Subtask;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.function.ThrowingFunction;

@Service
@RequiredArgsConstructor
@Slf4j
public class StoreService implements StoreUseCase {

  private final VirtualThreadManager virtualThreadManager = new VirtualThreadManager();

  private final StoreRepositoryPort storeRepositoryPort;
  private final StoreApplicationMapper mapper;
  private final StoreValidator validator;
  private final MessageSenderPort messageSenderPort;
  private final ReviewPort reviewPort;
  private final ProductPort productPort;
  private final StoreSearchRepositoryPort storeSearchRepositoryPort;

  @Transactional
  public void addStore(AddStoreCommand.Request req) {

    // 영업 시작 시간이 영업 종료 시간보다 작은지 확인
    validator.validateOpenTimeBeforeCloseTime(req.getOpenTime(), req.getCloseTime());

    Store store = mapper.addStoreToDomain(req);
    storeRepositoryPort.save(store);
  }

  @Transactional(readOnly = true)
  public DeliveryInfoCommand.Response getDeliveryInfo(DeliveryInfoCommand.Request req) {

    Store store = storeRepositoryPort.findById(req.getStoreId());
    return mapper.getDeliveryInfoToResponse(store);
  }

  @Transactional(readOnly = true)
  public void validateStoreOrderInfo(ValidateOrderInfoCommand.Request req) {

    try {
      Store store = storeRepositoryPort.findById(req.getStoreId());
      validator.validateDeliveryAmount(store, req.getDeliveryAmount());
      messageSenderPort.sendSuccessOrderValidate(req.getOrderTransactionId());
    } catch (Exception e) {
      messageSenderPort.sendFailOrderValidate(req.getOrderTransactionId(), e.getMessage());
    }
  }

  @Transactional(readOnly = true)
  public GetStoreDetailCommand getStoreDetail(Long storeId) {
    ThrowingFunction<ShutdownOnFailure, GetStoreDetailCommand> function = (ShutdownOnFailure scope) -> {
      Subtask<Store> store = scope.fork(() ->
          storeRepositoryPort.findById(storeId));

      Subtask<List<StoreReviewSummary>> top10Reviews = scope.fork(
          () -> reviewPort.getTop10Reviews(storeId));

      Subtask<Double> averageScore = scope.fork(() ->
          reviewPort.getAverageScore(storeId));

      Subtask<List<ProductInfo>> products = scope.fork(() ->
          productPort.findProductsByStoreId(storeId));

      scope.join()
          .throwIfFailed();

      GetStoreDetailCommand result = mapper.getStoreDetailToResponse(
          store.get(),
          top10Reviews.get(),
          averageScore.get(),
          products.get()
      );

      return result;
    };

    return StructuredTaskUtil.shutdownOnFailure(function);
  }

  @Override
  public List<Response> getSearchName(Request req) {
    List<Store> store = storeSearchRepositoryPort.findByNameContaining(req.getName());
    return mapper.searchNameToResponse(store);
  }
}

File: src/main/java/com/baedal/store/kotlin/application/service/StoreService.kt
package com.baedal.store.kotlin.application.service

import com.baedal.store.java.application.mapper.StoreApplicationMapper
import com.baedal.store.kotlin.application.command.GetStoreDetailCommand
import com.baedal.store.kotlin.application.port.`in`.StoreUseCase
import com.baedal.store.kotlin.application.port.out.ProductPort
import com.baedal.store.kotlin.application.port.out.ReviewPort
import com.baedal.store.kotlin.application.port.out.StoreRepositoryPort
import com.baedal.store.kotlin.domain.model.ProductInfo
import com.baedal.store.kotlin.domain.model.Store
import com.baedal.store.kotlin.domain.model.StoreDetailReview
import com.baedal.store.kotlin.domain.model.StoreReviewSummary
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional


/**
 *
 * note.
 *
 * Kotlin 의 클래스는, 기본적으로 Final 로 정의가 되기 때문에 Proxy 등을 생성하는
 * @Transactional 등을 활용할 경우, Open 키워드를 활용해야만 한다.
 *
 * Kotlin-Spring Plugin 을 활용하면 이러한 과정을 자동으로 처리해줘, 생략할 수 있도록 도와준다.
 */
//
@Service("kotlinStoreService")
open class StoreService (
    private val storeRepositoryPort: StoreRepositoryPort,
    private val storeMapper: StoreApplicationMapper,
    private val reviewPort: ReviewPort,
    private val productPort: ProductPort
): StoreUseCase {


    /**
     * note.
     *
     * 기존 ReviewScore 은, 1~5 사이의 정수를 강제하도록 Service 단에서 Enum 형식으로 데이터를 변환하고,
     * 이를 검증하였으나 어노테이션을 활용해 Controller 단에서 입력되지 않도록 방지하는 방안은 어떨지?
     * 만약, 위와 같이 변경하였다면 Service 단에서 2차 검증을 진행할 필요가 있을지?
     *
     */
    //
    @Transactional(readOnly = true)
    override fun getStoreDetail(storeId: Long): GetStoreDetailCommand {

        val store: Store = storeRepositoryPort.findById(storeId)

        val reviewInfo: StoreDetailReview = reviewPort.storeDetailReview(storeId)
        val top10Reviews: List<StoreReviewSummary> = reviewInfo.top10Reviews
        val averageScore: Double = reviewInfo.averageScore

        val products: List<ProductInfo> = productPort.findProductsByStoreId(storeId)

        return storeMapper.toCommand(store, top10Reviews, averageScore, products)
    }

}